### PR TITLE
validate CSMS-URL: opt. schema must fit with security-profile; chargepoint-ID deprecated in URL

### PR DESCRIPTION
**Pull request by Dominik-K**

### CODE REVIEW

**Comment by Pietfried** [comment]
So far it was required to include the chargepoint_id in the configured CentralSystemURI (v16) or ocppCsmsUrl (v201, as part of the NetworkConnectionProfiles variable). This is indeed redundant so I like this initiative!

For v16 we can append the internal configuration key ChargePointId (like implemented in this PR) and for v201 we need to use the variable "Identity" from the "SecurityCtrlr", so this line is currently not correct:
https://github.com/EVerest/libocpp/blob/main/lib/ocpp/v201/charge_point.cpp#L585

Please add this to the PR.

File: include/ocpp/helpers/uri.hpp
Comment (Pietfried) [inline include/ocpp/helpers/uri.hpp:1]: We already have a header and source file utils.hpp and utils.cpp in the common namespace. I would put this function there

File: include/ocpp/helpers/uri.hpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 Pionix GmbH and Contributors to EVerest
+#ifndef HELPERS_URI_HPP
+#define HELPERS_URI_HPP
+
+#include <string>
+
+namespace helpers {
+
+    std::string URIAppendPath(std::string uri, std::string path);
```
Comment (Pietfried) [inline include/ocpp/helpers/uri.hpp:10]: we decided to use snake_case for everything we add. Libocpp needs a general refactor to make this consistent

File: lib/ocpp/common/websocket/websocket_tls.cpp [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 // Copyright 2020 - 2023 Pionix GmbH and Contributors to EVerest
-#include <boost/optional/optional.hpp>
-
 #include <everest/logging.hpp>
 
 #include <ocpp/common/pki_handler.hpp>
 #include <ocpp/common/websocket/websocket_tls.hpp>
+#include <ocpp/helpers/uri.hpp>
+
+#include <boost/optional/optional.hpp>
```
Comment (Pietfried) [inline lib/ocpp/common/websocket/websocket_tls.cpp:9]: where do we need this header? we moved to std::optional a while ago

**Review by Pietfried — CHANGES_REQUESTED** [review]

File: lib/ocpp/common/websocket/websocket_tls.cpp (hunk 2 above)
Comment (Dominik-K) [reply lib/ocpp/common/websocket/websocket_tls.cpp:9]: Not needed. I guess it was included somehow by the IDE.

**Comment by Dominik-K** [comment]
Next I'll use a `Uri::set_path` to allow backwards-compatibility with current URLs which contain the CP-ID as path.

File: include/ocpp/common/websocket/websocket_uri.hpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#include <string>
+#include <websocketpp/uri.hpp>
+
+namespace ocpp {
+
+class Uri {
+public:
+    Uri(){};
+
+    void init(bool secure, std::string host, int port, std::string path); // TODO should return an error if invalid
```
Comment (Dominik-K) [inline include/ocpp/common/websocket/websocket_uri.hpp:16]: I think that's now unused.

File: lib/ocpp/common/websocket/websocket_plain.cpp [hunk 4]
```diff
@@ -10,27 +11,35 @@ namespace ocpp {
...
     this->ws_client.clear_error_channels(websocketpp::log::elevel::all);
     this->ws_client.init_asio();
     this->ws_client.start_perpetual();
-    this->uri = uri;
 
     websocket_thread.reset(new websocketpp::lib::thread(&client::run, &this->ws_client));
 
     this->reconnect_callback = [this](const websocketpp::lib::error_code& ec) {
-        EVLOG_info << "Reconnecting to plain websocket at uri: " << this->connection_options.cs_uri
+        EVLOG_info << "Reconnecting to plain websocket at uri: " << this->connection_options.csms_uri
```
Comment (Dominik-K) [inline lib/ocpp/common/websocket/websocket_plain.cpp:42]: Use `this->uri.string()`

**Comment by Dominik-K** [comment]
@Pietfried It's ready now. I'll do the returning an error/`std::expected` in a follow-up PR.

**Review by Pietfried — CHANGES_REQUESTED** [review]

File: lib/ocpp/common/websocket/websocket_base.cpp [hunk 5]
```diff
@@ -17,7 +16,10 @@ WebsocketBase::WebsocketBase(const WebsocketConnectionOptions& connection_option
     reconnect_backoff_ms(0),
     shutting_down(false),
     reconnecting(false) {
-    this->ping_timer = std::make_unique<Everest::SteadyTimer>();
+
+    set_connection_options_base(connection_options)
```
[the remark this answers is not in the document]
Comment (Dominik-K) [reply lib/ocpp/common/websocket/websocket_base.cpp]: Yes, sorry, I've forgot to re-build before committing, obviously.

File: lib/ocpp/common/websocket/websocket_uri.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    if (pos_last_slash == path.length() - 1) {
+        path.pop_back();
+        pos_last_slash -= 1;
+    }
+
+    auto path_without_last_segment = std::string(path.substr(0, pos_last_slash + 1));
+    std::string last_segment = std::string(path.substr(pos_last_slash + 1));
+
+    auto last_segment = std::string(path.substr(pos_last_slash + 1));
```
Comment (maaikez) [inline lib/ocpp/common/websocket/websocket_uri.cpp:35]: you have `last_segment` twice here, which does not compile. Please fix ( I don't know how to request a change but this definitely is :))

File: lib/ocpp/common/websocket/websocket_plain.cpp [hunk 7]
```diff
@@ -1,37 +1,57 @@
...
+    switch (connection_options.security_profile) { // `switch` used to lint on missing enum-values
+    case security::SecurityProfile::UNSECURED_TRANSPORT_WITH_BASIC_AUTHENTICATION:
+        break;
+    case security::SecurityProfile::TLS_WITH_BASIC_AUTHENTICATION:
+    case security::SecurityProfile::TLS_WITH_CLIENT_SIDE_CERTIFICATES:
+        throw std::invalid_argument("`security_profile` is not a plain, unsecured one.");
+    default:
+        throw std::invalid_argument("unknown `security_profile`, value = " +
+                                    std::to_string(connection_options.security_profile));
+    }
```
Comment (Dominik-K) [inline lib/ocpp/common/websocket/websocket_plain.cpp:31]: Note: the unofficial security-profile `0` is left out here

File: lib/ocpp/common/websocket/websocket_uri.cpp (hunk 6 above)
Comment (Dominik-K) [reply lib/ocpp/common/websocket/websocket_uri.cpp:35]: Yeah, sorry. I don't have any automated build running locally when committing stuff. Seems I've forgot it here.
But we're on adding build checks for all PR in CI :sweat_smile:

File: lib/ocpp/v16/charge_point_impl.cpp [hunk 8]
```diff
@@ -3446,4 +3449,4 @@ ConfigurationStatus ChargePointImpl::set_custom_configuration_key(CiString<50> k
 }
 
 } // namespace v16
-} // namespace ocpp
+} // namespace ocpp
```
Comment (Pietfried) [inline lib/ocpp/v16/charge_point_impl.cpp:3452]: missing newline

File: lib/ocpp/v201/charge_point.cpp [hunk 9]
```diff
@@ -679,19 +680,17 @@ WebsocketConnectionOptions ChargePoint::get_ws_connection_options(const int32_t
...
 
-    if (ocpp_csms_url.compare(0, 5, "ws://") == 0) {
-        ocpp_csms_url.erase(0, 5);
-    } else if (ocpp_csms_url.compare(0, 6, "wss://") == 0) {
-        ocpp_csms_url.erase(0, 6);
-    }
+    auto security_profile = network_connection_profile.securityProfile;
+    auto uri = Uri::parse_and_validate(
+        network_connection_profile.ocppCsmsUrl.get(),
+        this->device_model->get_value<std::string>(ControllerComponentVariables::SecurityCtrlrIdentity),
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:687]: Identity is defined as not required in OCPP2.0.1 spec (so it should be requested from storage with get_optional_value). This seems to be a problem with the spec. Lets define Identity as requried in [config/](https://github.com/EVerest/libocpp/blob/main/config/v201/component_schemas/standardized/SecurityCtrlr.json#L174) and keep requesting storage with get_value

File: lib/ocpp/v201/charge_point.cpp [hunk 10]
```diff
@@ -2423,4 +2422,4 @@ void ChargePoint::handle_get_local_authorization_list_version_req(Call<GetLocalL
 }
 
 } // namespace v201
-} // namespace ocpp
+} // namespace ocpp
```
Comment (Pietfried) [inline lib/ocpp/v201/charge_point.cpp:2425]: missing newline

File: include/ocpp/common/types.hpp [hunk 11]
```diff
@@ -558,6 +558,17 @@ firmware_status_notification_to_firmware_status_enum_type(const FirmwareStatusNo
 
 } // namespace conversions
 
+namespace security {
+// The security profiles defined in OCPP 2.0.1 resp. in the OCPP 1.6 security-whitepaper.
+enum SecurityProfile { // no "enum class" because values are used in implicit `switch`-comparisons to `int
+                       // security_profile`
+    UNSPECIFIED = 0,
```
Comment (Pietfried) [inline include/ocpp/common/types.hpp:565]: Rather:
UNSECURED_TRANSPORT_WITHOUT_BASIC_AUTHENTICATION

File: lib/ocpp/common/websocket/websocket_plain.cpp (hunk 7 above)
Comment (Pietfried) [reply lib/ocpp/common/websocket/websocket_plain.cpp:31]: This should not be left out here

**Review by Pietfried — CHANGES_REQUESTED** [review]

File: include/ocpp/common/types.hpp (hunk 11 above)
Comment (Dominik-K) [reply include/ocpp/common/types.hpp:565]: I've added the prefix `OCPP_1_6_` to clearly indicate it's only possible within 0CPP1.6.

File: lib/ocpp/v201/charge_point.cpp (hunk 9 above)
Comment (Dominik-K) [reply lib/ocpp/v201/charge_point.cpp:687]: Is there sth. more needed except adding `Identity` to `required:` in `SecurityCtrlr.json`?

**Review by Pietfried — APPROVED** [review]

**Pull request merged**